python/ray/serve/async_goal_manager.py

```diff
@@ -8,34 +8,48 @@
 
 
 class AsyncGoalManager:
+    """
+    Class to help ServeController to manage desirable states of replica.
+
+    Essentially a dictionary of <goal_id, asyncio.Future> within
+    ServeController's event loop with APIs to facilitate common calls.
+    """
+
     def __init__(self):
-        self._pending_goals: Dict[GoalId, asyncio.Event] = dict()
+        self._pending_goals: Dict[GoalId, asyncio.Future] = dict()
 
     def num_pending_goals(self) -> int:
         return len(self._pending_goals)
 
     def create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
         if goal_id is None:
             goal_id = uuid4()
-        self._pending_goals[goal_id] = asyncio.Event()
+
+        self._pending_goals[
+            goal_id] = asyncio.get_running_loop().create_future()
         return goal_id
 
     def complete_goal(self, goal_id: GoalId) -> None:
         logger.debug(f"Completing goal {goal_id}")
-        event = self._pending_goals.pop(goal_id, None)
-        if event:
-            event.set()
+        future = self._pending_goals.pop(goal_id, None)
+        if future:
+            future.set_result(goal_id)
```

Comment:
and move set_exception here

Reply:
+1, I think it's better to have a single place to complete a goal. Maybe we should also let the caller pass in the right exception object instead of using a bare `Exception`?